kong/tools/wrpc/threads.lua

```diff
@@ -0,0 +1,99 @@
+local pb = require "pb"
+local utils = require "kong.tools.wrpc.utils"
+local message = require "kong.tools.wrpc.message"
+
+local safe_args = utils.safe_args
+local endswith = utils.endswith
+local pb_decode = pb.decode
+
+local ngx_now = ngx.now
+local ngx_log = ngx.log
+local NOTICE = ngx.NOTICE
+local exiting = ngx.worker.exiting
+local sleep = ngx.sleep
+local thread_spawn = ngx.thread.spawn
+local thread_kill = ngx.thread.kill
+local thread_wait = ngx.thread.wait
+
+local process_message = message.process_message
+local handle_error = message.handle_error
+
+local _M = {}
+
+local function step(wrpc_peer)
+  local msg, err = wrpc_peer:receive()
+
+  while msg ~= nil do
+    msg = assert(pb_decode("wrpc.WebsocketPayload", msg))
+    assert(msg.version == "PAYLOAD_VERSION_V1", "unknown encoding version")
+    local payload = msg.payload
+
+    if payload.mtype == "MESSAGE_TYPE_ERROR" then
+      handle_error(wrpc_peer, payload)
+
+    elseif payload.mtype == "MESSAGE_TYPE_RPC" then
+      local ack = payload.ack or 0
+      local deadline = payload.deadline or 0
+
+      if ack == 0 and deadline < ngx_now() then
+        ngx_log(NOTICE, "[wRPC] Expired message (", deadline, "<", ngx_now(), ") discarded")
+
+      elseif ack ~= 0 and deadline ~= 0 then
+        ngx_log(NOTICE, "[WRPC] Invalid deadline (", deadline, ") for response")
```

Comment:
i think the logic is correct and more readable, but depending on some decoding flags, those fields (`ack` and `deadline`) could arrive here with `0` as the "undefined" value, and thus not falsy.  that's because of the "not exactly static-type" semantics of protobuf.

Reply:
I will leave it unchanged but add comments to explain this.